core/observe.c

```diff
@@ -964,39 +941,40 @@ static void prv_obsRequestCallback(lwm2m_context_t * contextP,
-                                  code,  //?
-                                  NULL,
-                                  utils_convertMediaType(packet->content_type),
-                                  packet->payload,
-                                  packet->payload_len,
-                                  observationData->userData);
-    }
-    else
-    {
-        if(observationP == NULL)
-        {
+    if (code != COAP_205_CONTENT) {
+        // Some kind of error occurred, call the request callback with an error code
+        observationData->callback(contextP, observationData->client, &observationData->uri,
+                                  code, //?
+                                  NULL, LWM2M_CONTENT_TEXT, NULL, 0, observationData->userData);
+    } else if (IS_OPTION(packet, COAP_OPTION_BLOCK2) && packet->block2_more) {
+        // Call request callback with partial block2 content.
+        observationData->callback(contextP, observationData->client, &observationData->uri,
+                                  code, //?
+                                  NULL, utils_convertMediaType(packet->content_type), packet->payload,
+                                  packet->payload_len, observationData->userData);
+    } else {
+        int has_block2 = coap_get_header_block2(packet, &block_num, &block_more, &block_size, NULL);
+        if (has_block2) {
+            block_info.block_num = block_num;
+            block_info.block_size = block_size;
+            block_info.block_more = block_more;
+        }
+
+        if (observationP == NULL) {
             observationP = (lwm2m_observation_t *)lwm2m_malloc(sizeof(*observationP));
-            if (observationP == NULL) goto end;
+            if (observationP == NULL) {
+                transaction_free_userData(contextP, transacP);
+                return;
+            }
             memset(observationP, 0, sizeof(*observationP));
-        }
-        else
-        {
-            observationP->clientP->observationList = (lwm2m_observation_t *) LWM2M_LIST_RM(observationP->clientP->observationList, observationP->id, NULL);
+        } else {
+            observationP->clientP->observationList =
+                (lwm2m_observation_t *)LWM2M_LIST_RM(observationP->clientP->observationList, observationP->id, NULL);
 
             // give the user chance to free previous observation userData
             // indicator: COAP_202_DELETED and (Length ==0)
-            observationData->callback(contextP,
-                                      observationData->client,
-                                      &observationData->uri,
-                                      COAP_202_DELETED,
-                                      NULL,
-                                      LWM2M_CONTENT_TEXT, NULL, 0,
-                                      observationData->userData);
+            observationData->callback(contextP, observationData->client, &observationData->uri, COAP_202_DELETED, NULL,
```

Comment:
Is there point to run callback with new observation data?

Mayhaps is better to run old callback of found observation with it's data?
```
observationP->callback(contextP, observationP->clientP->internalID, &observationP->uri,
                             COAP_202_DELETED,
                             NULL, LWM2M_CONTENT_TEXT, NULL, 0, observationP->userData);
```

I understand that yours change was only formatting, anyway

Reply:
Yes, this is formatting only as a result of running clang-format. Frankly, I am not a 100% sure how to reach this code. It looks like the callback is triggered in an edge case to be able to free observation `userData` in the callback, if you provided a pointer to `userData` on the heap when starting the observation with `lwm2m_observe(...)` (however, this is not used by the server example). I updated the PR to remove the "formatting only" changes to keep the history a bit cleaner.

Reply:
You can reach it by sending second obervation request for the same object.
New observation replaces the old one and old callback will not be called for every new observed value.

I assume previous userData should be freed in observe callback:
```
if (dataLength == 0 && status == COAP_202_DELETED) {
    free((mystruct_t *)userData);
}
```

Currently, if I do this, fresh memory allocation is destroyed, not the old on.

This is not related to #619, should I open new issue?

Thanks.

Reply:
Yes, we should track that in a new issue, so this can be closed.